expertsystem/data.py

```diff
@@ -84,29 +99,121 @@ def __hash__(self) -> int:
         return hash(repr(self))
 
 
-class Particle(NamedTuple):
-    """Immutable data container for particle info."""
+@dataclass(frozen=True)
+class HasQuantumNumbers:  # pylint: disable=too-many-instance-attributes
+    """Set of quantum numbers **excluding spin**.
 
-    name: str
-    pid: int
-    charge: int
-    spin: float
-    mass: float
+    This is to make spin projection required in `.QuantumState` (`.HasSpin`)
+    and unavailable in `.Particle` (`.HasFloatSpin`).
+    """
+
+    charge: int = 0
+    isospin: Optional[Spin] = None
     strangeness: int = 0
     charmness: int = 0
     bottomness: int = 0
     topness: int = 0
     baryon_number: int = 0
-    electron_number: int = 0
-    muon_number: int = 0
-    tau_number: int = 0
-    width: Optional[float] = None
-    isospin: Optional[Spin] = None
+    electron_lepton_number: int = 0
+    muon_lepton_number: int = 0
+    tau_lepton_number: int = 0
     parity: Optional[Parity] = None
     c_parity: Optional[Parity] = None
     g_parity: Optional[Parity] = None
 
 
+@dataclass(frozen=True)
+class HasSpin:
+    """Required to disallow default arguments for `.QuantumState`."""
+
+    spin: Spin
+
+
+@dataclass(frozen=True)
+class HasFloatSpin:
+    """Required to disallow default arguments for `.ParticleQuantumState`."""
+
+    spin: float
+
+
+@dataclass(frozen=True)
+class QuantumState(HasQuantumNumbers, HasSpin):
+    """Contains all quantum numbers unambiguously defining a quantum state."""
+
+
+@dataclass(frozen=True)
+class ParticleQuantumState(HasQuantumNumbers, HasFloatSpin):
+    """Similar to `.QuantumState` but only carrying spin magnitude."""
+
+
+class HasComplexEnergy:
```

Comment:
Best add `__eq__` here and just call it from the parent classes.

Reply:
https://github.com/ComPWA/expertsystem/pull/135#discussion_r464848162